### PR TITLE
Permission update from Directory.Read.All to GroupMember.Read.All

### PR DESCRIPTION
**Pull request by Shama-K**
## Purpose
Updated code and Readme files to include GroupMember.Read.All permission. Updated implementation to handle Groups overage scenario.

Note: Code snippet will be updated once code is approved.

## Does this introduce a breaking change?

```
[ ] Yes
[x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ x] Refactoring (no functional changes, no api changes)
[ x] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
Use manual steps or app creation scripts to configure the sample, and run the sample.
Test the code.

### CODE REVIEW

File: 5-WebApp-AuthZ/5-2-Groups/README.md [hunk 1]
```diff
@@ -10,10 +10,11 @@ page_type: sample
 languages:
   - csharp  
 products:
-  - azure
+  - microsoft-authentication-library
+  - microsoft-identity-platform					   
   - azure-active-directory  
   - dotnet
-  - office-ms-graph
+  - microsoft-graph-api
```
Comment (derisen) [inline 5-WebApp-AuthZ/5-2-Groups/README.md:17]: "ms-graph" is the correct label

File: 5-WebApp-AuthZ/5-2-Groups/README-incremental-instructions.md [hunk 2]
```diff
@@ -10,10 +10,11 @@ page_type: sample
 languages:
   - csharp  
 products:
-  - azure
+  - microsoft-authentication-library
+  - microsoft-identity-platform		   
   - azure-active-directory  
   - dotnet
-  - office-ms-graph
+  - microsoft-graph-api
```
Comment (derisen) [inline 5-WebApp-AuthZ/5-2-Groups/README-incremental-instructions.md:17]: "ms-graph"

File: 5-WebApp-AuthZ/5-2-Groups/README.md [hunk 3]
```diff
@@ -99,20 +96,18 @@ There is one project in this sample. To register it, you can:
...
 
 As a first step you'll need to:
 
-1. Sign in to the [Azure portal](https://portal.azure.com) using either a work or school account or a personal Microsoft account.
-1. If your account is present in more than one Azure AD tenant, select your profile at the top right corner in the menu on top of the page. Then select **switch directory** to change your portal session to the desired Azure AD tenant.
+1. Sign in to the [Azure portal](https://portal.azure.com).
+1. If your account is present in more than one Azure AD tenant, select your profile at the top right corner in the menu on top of the page, and then **switch directory**.
 
-#### Register the webApp app (WebApp-GroupClaims)
+#### Register the Web App (WebApp-GroupClaims)
```
Comment (jmprieur) [inline 5-WebApp-AuthZ/5-2-Groups/README.md:110]: I think Marsh recommended "web app".

File: 5-WebApp-AuthZ/5-2-Groups/Infrastructure/CustomAuthorization.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Filters;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+namespace WebApp_OpenIDConnect_DotNet.Infrastructure
+{
+    public class GroupPolicyHandler : AuthorizationHandler<GroupPolicyRequirement>
```
Comment (kalyankrishna1) [inline 5-WebApp-AuthZ/5-2-Groups/Infrastructure/CustomAuthorization.cs:10]: >GroupPolicyHandler [](start = 17, length = 18)

Add comments describing the purpose of the class

**Review by derisen — APPROVED** [review]
Setup, registrations and running steps are clear. App works without issue. LGTM -thanks Shama

File: 5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+namespace WebApp_OpenIDConnect_DotNet.Services
+{
+    public class GraphHelper
+    {
+        /// <summary>
+        /// This method inspects the claims collection created from the ID or Access token issued to a user and returns the groups that are present in the token . If it detects groups overage,
+        /// the method then makes calls to Microsoft Graph to fetch the group membership of the authenticated user.
+        /// </summary>
+        /// <param name="context">TokenValidatedContext</param>
+        public static async Task<List<string>> GetSignedInUsersGroups(TokenValidatedContext context)
```
Comment (kalyankrishna1) [inline 5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs:21]: >GetSignedInUsersGroups [](start = 47, length = 22)

1. check if token overgae occured, if yes
 1.1 check for the existance of the session var ( if (_httpContextAccessor.HttpContext.Session.Keys.Contains("groupClaims")))
  1.1.1 if session var the groups of the user fetched earlier, return the groups
  1.1.1 if session var does not have the groups of the user, fetch the groups from graph and populate the session var
2. return the groups array for the caller

File: 5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using WebApp_OpenIDConnect_DotNet.Infrastructure;
+
+namespace WebApp_OpenIDConnect_DotNet.Services
+{
+    public class GraphHelper
```
Comment (kalyankrishna1) [inline 5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs:14]: >GraphHelper [](start = 17, length = 11)

1. ProcessUserGroupsForOverage
2. CheckUsersGroupMembership

File: 5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        /// <param name="context">TokenValidatedContext</param>
+        public static async Task<List<string>> GetSignedInUsersGroups(TokenValidatedContext context)
+        {
+            List<string> groupClaims = new List<string>();
+
+            try
+            {
+                // Checks if the incoming token contained a 'Group Overage' claim.
+                if (context.Principal.Claims.Any(x => x.Type == "hasgroups" || (x.Type == "_claim_names" && x.Value == "{\"groups\":\"src1\"}")))
+                {
```
Comment (kalyankrishna1) [inline 5-WebApp-AuthZ/5-2-Groups/Services/MicrosoftGraph-Rest/GraphHelper.cs:29]: bool HasOverageOccurred

File: 1-WebApp-OIDC/1-1-MyOrg/AppCreationScripts/createdApps.html [hunk 8]
```diff
@@ -0,0 +1,4 @@
+<html><body><table>
+<thead><tr><th>Application</th><th>AppId</th><th>Url in the Azure portal</th></tr></thead><tbody>
+<tr><td>webApp</td><td>533843a1-3f30-4e5b-8800-7d974c29837e</td><td><a href='https://portal.azure.com/#blade/Microsoft_AAD_RegisteredApps/ApplicationMenuBlade/CallAnAPI/appId/533843a1-3f30-4e5b-8800-7d974c29837e/objectId/fe605bc2-667f-4f18-b534-eb243eb686b1/isMSAApp/'>WebApp</a></td></tr>
```
Comment (kalyankrishna1) [inline 1-WebApp-OIDC/1-1-MyOrg/AppCreationScripts/createdApps.html:3]: delete this please

File: 4-WebApp-your-API/4-3-AnyOrg/ToDoListClient/appsettings.json [hunk 9]
```diff
@@ -1,10 +1,10 @@
 {
   "AzureAd": {
     "Instance": "https://login.microsoftonline.com/",
-    "Domain": "[Enter the domain of your tenant, e.g. contoso.onmicrosoft.com]",
+    "Domain": "msFreeTenant.onmicrosoft.com",
     "TenantId": "common",
-    "ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
-    "ClientSecret": "[Copy the client secret added to the app from the Azure portal]",
+    "ClientId": "176b110b-80ce-4943-8fb7-3b19fe210d46",
+    "ClientSecret": "3W/iqNToW00tW7bt358jPMNXeZ4Icn/w9QxjP8ZWQxA=",
```
Comment (kalyankrishna1) [inline 4-WebApp-your-API/4-3-AnyOrg/ToDoListClient/appsettings.json:7]: >"ClientSecret": "3W/iqNToW00tW7bt358jPMNXeZ4Icn/w9QxjP8ZWQxA=", [](start = 4, length = 63)

please remove this and revert

File: 4-WebApp-your-API/4-3-AnyOrg/TodoListService/appsettings.json [hunk 10]
```diff
@@ -1,10 +1,10 @@
 {
   "AzureAd": {
     "Instance": "https://login.microsoftonline.com/",
-    "Domain": "[Enter the domain of your tenant, e.g. contoso.onmicrosoft.com]",
+    "Domain": "msFreeTenant.onmicrosoft.com",
     "TenantId": "common",
-    "ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
-    "ClientSecret": "[Copy the client secret added to the app from the Azure portal]",
+    "ClientId": "0892ebc9-a57e-4b64-814f-4bd0798f8132",
+    "ClientSecret": "rZ0TVcb901KRnDQ4TEcvdPuxr2ABimk80Mg4nV8Tl4o=",
```
Comment (kalyankrishna1) [inline 4-WebApp-your-API/4-3-AnyOrg/TodoListService/appsettings.json:7]: >Z0TVcb901KRnDQ4TEcvdPuxr2ABimk80Mg4nV8Tl4o [](start = 22, length = 42)

please revert this

File: 5-WebApp-AuthZ/5-1-Roles/appsettings.json [hunk 11]
```diff
@@ -1,14 +1,14 @@
...
-    "ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
+    "Domain": "ms0604.onmicrosoft.com",
+    "TenantId": "d963c147-dc13-433f-9520-6db1ff177c34",
+    "ClientId": "42d382ca-83ad-4384-b7ca-c732d90be3f2",
     "CallbackPath": "/signin-oidc",
     "SignedOutCallbackPath ": "/signout-callback-oidc",
 
     // To call an API
-    "ClientSecret": "[Copy the client secret added to the app from the Azure portal]"
+    "ClientSecret": "Ne_h-tm_5oT6k5B_IM17_WNp.RItqTb.-E"
```
Comment (kalyankrishna1) [inline 5-WebApp-AuthZ/5-1-Roles/appsettings.json:11]: >Secret": "Ne_h-tm_5oT6k5B_IM17_WNp.RItqTb.-E [](start = 11, length = 44)

please revert this

File: 4-WebApp-your-API/4-3-AnyOrg/ToDoListClient/appsettings.json (hunk 9 above)
Comment (Shama-K) [reply 4-WebApp-your-API/4-3-AnyOrg/ToDoListClient/appsettings.json]: Reverted.

File: 5-WebApp-AuthZ/5-1-Roles/appsettings.json (hunk 11 above)
Comment (Shama-K) [reply 5-WebApp-AuthZ/5-1-Roles/appsettings.json]: Reverted.
Thanks.

**Review by TiagoBrenck — APPROVED** [review]
@LGTM, thanks Shama

File: 5-WebApp-AuthZ/5-2-Groups/README.md [hunk 12]
```diff
@@ -342,59 +358,223 @@ The following files have the code that would be of interest to you:
...
+              }
+              return groupClaims;
+          }
+      ```
 
-      services.AddMSGraphService(Configuration);    // Adds the IMSGraphService as an available service for this app.
+    **ProcessClaimsForGroupsOverage** method uses `GraphServiceClient` to retrieve groups for the signed-in user from [/me/memberOf](https://docs.microsoft.com/graph/api/user-list-memberof) endpoint. All the group ids are stored in Session.
+
+      ```csharp
+          private static async Task<List<string>> ProcessUserGroupsForOverage(TokenValidatedContext context)
```
Comment (kalyankrishna1) [inline 5-WebApp-AuthZ/5-2-Groups/README.md:429]: >ProcessUserGroupsForOverage [](start = 50, length = 27)

you don't need to document his method

Comment (Shama-K) [reply 5-WebApp-AuthZ/5-2-Groups/README.md]: Updated the Readme files.
Thanks.

**Review by kalyankrishna1 — APPROVED** [review]
:shipit:

**Pull request merged**